docs/rfc/active/5818-serial-updates.md

```diff
@@ -0,0 +1,313 @@
+
+### 4: Serialised Updates Exclusively via Kafka
+
+Here, the client never updates the aspect HTTP but instead passes it via the MetadataChangeProposal topic in Kafka. 
+I don't think this is a viable solution due to the possibility of issuing out-of-date updates.
+
+Good
+* Local ordering
+* Fast performance
+
+Bad
+* No way of knowing what the current state is unless the client constructs it from the Kafka topics
+* Danger of proposing out-of-date updates if the topic is lagging
+
+### 5: URN-Aspect Locking
+
+Lock the target URN aspect(s) when reading. The client is guaranteed that the state it has read will not change until it releases the lock.
+
+Good
+* Guarantee of state between the read and write
+* No need for business-specific logic in GMS code
+
+Bad
+* Slow, especially under high contention
+* Need to handle lock release if a client fails to release the lock
+* Potentially blocking MCE Kafka queue
+
+[![](https://mermaid.ink/img/pako:eNqNksFOwzAMhl8lygm07QV6GEIw7QDTUKuJSy9W4tGINimpc0DT3h136dZQjYkequj3p9-_nRykchplJjv8CmgVPhv48NCUVvAHipwXT7VBS1GJ58VyOVtvikyskR67FhW9G6penfqMFNd6JJZ6ORP9f-dtlFJqgwQaCApuhYljZH5VF4yf2qrgPceYmnE55psQs_qSbMifkG9AqhKQWE1m3LUcAKPTXe90fyN-Ct-aANqWu-iHYc81nZUoXJlo-xJLWHco0Hvn_0RXeb7NB9rqq0PlWCN0ON5QOlN6bwN4ubqRdS0JMg26QGOSEWKfxf-MOKOcywZ9A0bzUzz0cimpwgZLmfFR4x5CTaUs7ZFRCOSKb6tkRj7gXIbT0oeXK7M98IrO6kobXvwgHn8AOC71Xg)](https://mermaid.live/edit#pako:eNqNksFOwzAMhl8lygm07QV6GEIw7QDTUKuJSy9W4tGINimpc0DT3h136dZQjYkequj3p9-_nRykchplJjv8CmgVPhv48NCUVvAHipwXT7VBS1GJ58VyOVtvikyskR67FhW9G6penfqMFNd6JJZ6ORP9f-dtlFJqgwQaCApuhYljZH5VF4yf2qrgPceYmnE55psQs_qSbMifkG9AqhKQWE1m3LUcAKPTXe90fyN-Ct-aANqWu-iHYc81nZUoXJlo-xJLWHco0Hvn_0RXeb7NB9rqq0PlWCN0ON5QOlN6bwN4ubqRdS0JMg26QGOSEWKfxf-MOKOcywZ9A0bzUzz0cimpwgZLmfFR4x5CTaUs7ZFRCOSKb6tkRj7gXIbT0oeXK7M98IrO6kobXvwgHn8AOC71Xg)
+
+### 6: PATCH with JSON-Delta Format
+
+Client sends its update request in JSON-delta or some other delta format. GMS will reject the update if the previous 
+side of the delta does not match.
+
+## How we teach this
+
+> What names and terminology work best for these concepts and why? How is this idea best presented? As a continuation
+> of existing DataHub patterns, or as a wholly new one?
+
+> What audience or audiences would be impacted by this change? Just DataHub backend developers? Frontend developers?
+> Users of the DataHub application itself?
+
+> Would the acceptance of this proposal mean the DataHub guides must be re-organized or altered? Does it change how
+> DataHub is taught to new users at any level?
+
+> How should this feature be introduced and taught to existing audiences?
+
+We would have two potential additions to terminology:
+1. Previous state: The version(s) of aspects required in order for a GM update to succeed.
+2. Preconditions: A wider set of assertions which must be true in order for a GM update to succeed.
+
+We don't really cover "Preconditions" in this RFC, and in fact I argue we should never do, as this leaks business 
```

Comment:
Check out the term "etag" -- its used across other DBs to provide a "version tag" to a document. I'm hoping we can simply add something similar to all of our get / put pathways

Reply:
I like it. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag seems to cover roughly what we need, though we'd need to think about the multiple aspect use case.

Reply:
Even better article here, which describes almost exactly what we want: https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#avoiding_the_lost_update_problem_with_optimistic_locking